### DISCUSSION TITLE
Broken database or bad program building?

### DISCUSSION THREAD

**Discussion by LousisMT** [textbrowser/biblioteq#293] [Q&A] [answered]
Hi all
I've upgraded to 2023.03.15. Apparently the process of compiling, making release etc worked fine.
But everything's a trouble now. I can't add books, make searchings o populate my database properly. It's clear there's something wrong, but I'm far from being an expert in sqlite and surroundings.
Any idea?
Here they're some pictures.
Thnks a lot in advance.
![biblioteq_trouble_04](https://user-images.githubusercontent.com/132607219/236638059-eecdd24f-8507-49af-aa34-c8cc3663af4f.jpg)
![biblioteq_trouble_01](https://user-images.githubusercontent.com/132607219/236638062-58b1773d-0300-4be6-a6ff-a532252859eb.jpg)
![biblioteq_trouble_02](https://user-images.githubusercontent.com/132607219/236638064-c114ae01-8193-4393-9818-91225a378239.jpg)
![biblioteq_trouble_03](https://user-images.githubusercontent.com/132607219/236638067-219b70c0-da6d-4053-a187-606423162629.jpg)
![biblioteq_trouble_00](https://user-images.githubusercontent.com/132607219/236638069-a7c7da4a-b737-4a66-ac2e-54601a2cb467.jpg)

**Comment by textbrowser** [comment]
If you're compiling source from the modern repository and expecting that BQ should recognize your state of affairs, do not. :)

If you've downloaded a bundle, you may or may not see a request to upgrade your database. But! It's complicated. There are numerous, hundreds perhaps, database revisions. I archive those in the source. You can resurface them, recompile, and attempt to upgrade your database.

I don't know the state of your database.

Every release includes PostgreSQL information. The source also includes it for SQLite. However, it's not cumulative. If your last version was from 2020 and you're now utilizing 2023, well, you'll have 2023 and not everything from 2020 to 2023.

So, you'll have to review your database and upgrade it.

**Comment by textbrowser** [comment]
People have requested this similar problem several times. A good solution doesn't exist. I can include version information in SQLite and spend lots and lots of hours determining how your database should be upgraded. But why? You can do that by remaining current as well as through the SQLite interface. :)

**Reply by meteos77** [reply]
I also have broken bases, to recover them, you just have to re-apply the SQL Schema Upgrade?

**Reply by textbrowser** [reply]
The answer is not simple.

**Comment by textbrowser** [comment, marked as answer]
https://github.com/textbrowser/biblioteq/blob/master/SQL/postgresql_update_schema.sql

/*
** Please do not load this file in its entirety.
*/

**Comment by textbrowser** [comment]
A similar process exists for SQLite. BQ doesn't know your database. It shouldn't It's your data, right?

**Comment by textbrowser** [comment]
You know when you upgrade a Debian distribution, some files in /etc are sometimes changed. Debian provides you with a merge if it can. In the end, you have to accept the revisions. If it cannot resolve your local files, etc. Whatever the reason. BQ has a very convenient and free upgrade mechanism for both engines! It even asks you to perform the upgrade. But, it's for the current version versus whatever your database looks like to BQ. It's not going to review all 21 years of revisions. Like, what are you expecting?

**Comment by textbrowser** [comment]
Discussion is beyond my free time. :) Closing!

**Comment by textbrowser** [comment]
Every single release, every one of them, includes information for upgrading the database engines. The information is in the release notes, in the source, and in separate files. You can't escape it! You'd have to be willingly ignorant to! Cannot be done. I've tried it, numerous times. I've tried it. Like, how can I avoid upgrading a database? Well, you see, queries fail. So, I can continue failing or I can put in the effort to make them succeed. Cannot be done otherwise. There is plenty of information in the error logs too.

I know what you're asking. You're asking: why doesn't BQ upgrade my database automatically without me thinking about it? Well, it does. It reminds you to do it as long as you're not 1000 years lost. It reminds you to retain a copy in case things die. Everything's there.

**Comment by textbrowser** [comment]
https://github.com/textbrowser/biblioteq/blob/master/Source/biblioteq_b.cc

Lines 5107 to 5469 are all about upgrades. Those are a lot of upgrades. So they're archived because some of them are expensive and the expectation is that you need to remain current. But if you cannot for whatever reason, you can execute those by rebuilding BQ with some or all of them.

I am creating a text file similar to PQ for SQ people. It will include the most current upgrade statements.

**Comment by textbrowser** [comment]
I've also helped upgrade people's databases by doing this process for them. Free! Without payment. Like, here let me help you because my life is so dull and my time so unlimited and pointless, that I need to do this for you.

**Comment by textbrowser** [comment]
81 update statements!

**Comment by textbrowser** [comment]
In addition to all of this information, you may also create a new database and merge your existing database.

https://github.com/textbrowser/biblioteq/discussions/191

This avoids an upgrade.

**Comment by textbrowser** [comment]
And if you discover problems, you can report them to keep the merge process correct and helpful. So many solutions and information. Free!

**Comment by textbrowser** [comment]
Available in the repository.

![bq](https://user-images.githubusercontent.com/10701156/236648574-d9d8a2d3-4b64-40ed-bbd7-13715cb2ed26.png)

**Reply by meteos77** [reply]
Great, thanks a lot for this new feature

**Reply by meteos77** [reply]
You have made a great cleaning in the discussions: we see more clearly :-)

**Comment by LousisMT** [comment]
I beg your pardon.
Fixed by merging databases.
Thanks a lot.
(Btw, I do know your work and I do respect the value of your time. That's why I donated the first day I used BiblioteQ, years ago.)
